[nciplugin] Fix wrong compare logic. JB#51201

ikamaletdinov opened the pull request:
Comparison logic during reactivation was wrong for t2 and t4 tags.
This commit changes behaviour to desired.

monich commented:
`libncicore_version` in `rpm/libnciplugin.spec` needs to be bumped to 1.1.13

monich commented on src/nci_adapter.c:
```diff
@@ -67,6 +68,7 @@ typedef struct nci_adapter_intf_info {
     NCI_MODE mode;
     GUtilData mode_param;
     GUtilData activation_param;
+    NciModeParam* mode_param_parsed;                /* Since 1.0.11 */
```
No need for these Since tags in the internal headers/sources. They only make sense for public interfaces, which other binaries can link against (to help those other binaries to keep their package requirements up-to-date).

ikamaletdinov replied:
Thanks a lot for the review, comment was deleted

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    return prev->sel_res == curr->sel_res &&
+        prev->sel_res_len == curr->sel_res_len &&
+        !memcmp(prev->sens_res, curr->sens_res,
+            FIELD_SIZEOF(NciModeParamPollA, sens_res));
```
I would prefer just `sizeof(curr->sens_res)` or `sizeof(prev->sens_res)`. IMO it's easier to read and as a bonus, that would squeeze memcmp part of the statement into a single line.

ikamaletdinov replied:
Thanks a lot for the review, I have replaced that part of code

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    changed after losing field */
+    return prev->fsc == curr->fsc &&
+        !memcmp(prev->app_data, curr->app_data,
+            FIELD_SIZEOF(NciModeParamPollB, app_data)) &&
```
Same here

ikamaletdinov replied:
Thanks a lot for the review, I have replaced that part of code

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+            FIELD_SIZEOF(NciModeParamPollB, app_data)) &&
+        prev->prot_info.size == curr->prot_info.size &&
+        !memcmp(prev->prot_info.bytes, curr->prot_info.bytes,
+            curr->prot_info.size);
```
Use `gutil_data_equal(&prev->prot_info, &curr->prot_info)` for comparison. The function is defined in `<gutil_misc.h>`. Bump libglibutil requirement in `rpm/libnciplugin.spec` to 1.0.31 (for build and run time)

ikamaletdinov replied:
Thanks a lot for the review, I have added method from gutil_misc.h

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+}
+
+static
+inline
```
I don't think it makes sense to force inline for every function. Let compiler do its job. If the function is static and is only used once, most likely it's going to be inlined in release build anyway.

ikamaletdinov replied:
Sound reasonable, deleted

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    const NciModeParamPollA* prev,
+    const NciModeParamPollA* curr)
+{
+    gboolean partial_match =  mode_param_match_poll_a(prev, curr);
```
Space after '=' and no empty line after the declaration.

ikamaletdinov replied:
Thanks, fixed

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    gboolean partial_match =  mode_param_match_poll_a(prev, curr);
+    /* For tag type 2 logic is almost the same, but random UID has some
+    limitations: according to AN10927 Random UID RID should be handled
+    separately - single sized (4 bytes) starting with 0x08 */
```
Please format multi-line comments like this:
```c
    /*
     * foo
     * bar
     */
```

ikamaletdinov replied:
Thanks, fixed

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    /* For tag type 2 logic is almost the same, but random UID has some
+    limitations: according to AN10927 Random UID RID should be handled
+    separately - single sized (4 bytes) starting with 0x08 */
+    if(prev->nfcid1_len == curr->nfcid1_len &&
```
Missing space after `if`

ikamaletdinov replied:
Thanks, fixed

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+        curr->nfcid1[0] == RANDOM_UID_START_BYTE) {
+        return partial_match;
+    } else {
+    /* otherwise UID should fully match */
```
Align the comment with the next line, **C**apitalize the first word for consistency

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+            !memcmp(prev->nfcid1, curr->nfcid1, curr->nfcid1_len);
+
+    }
+
```
Empty lines (one more before the bracket)

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+{
+    const NciModeParam* mode_param_prev = info->mode_param_parsed;
+    const NciModeParam* mode_param_current = ntf->mode_param;
+    /* mode params criteria depends on type of tag */
```
Again, empty line after the declaration, Capitalization

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    const NciIntfActivationNtf* ntf)
+{
+    const NciModeParam* mode_param_prev = info->mode_param_parsed;
+    const NciModeParam* mode_param_current = ntf->mode_param;
```
Does it matter which one is previous and which one is current? I don't think it does, please rename those into `mp1` and `mp2` to make code more compact.

monich replied:
The same applies to other places where it doesn't matter what's previous and what's current.

ikamaletdinov replied:
Thanks a lot for the review, I have replaced that part of code

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    switch (ntf->mode) {
+    case NCI_MODE_PASSIVE_POLL_A:
+        switch (ntf->rf_intf) {
+        case NCI_RF_INTERFACE_FRAME:
+            /* Type 2 Tag */
+            if(mode_param_prev && mode_param_current) {
+                const NciModeParamPollA* poll_a_prev = &mode_param_prev->poll_a;
+                const NciModeParamPollA* poll_a_current =
+                    &mode_param_current->poll_a;
+
+                return mode_param_match_poll_a_t2(poll_a_prev, poll_a_current);
+            }
```
Could've been just
```c
        case NCI_RF_INTERFACE_FRAME: /* Type 2 Tag */
            return mode_param_match_poll_a_t2(&mp1->poll_a, &mp2->poll_a);
```
Let the caller make sure that pointers are not NULL (which it already does I think).

ikamaletdinov replied:
I have refactored that part of code. But unfortunately, parameters could be NULL, but I have moved that check one layer above (before switch)

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+                return mode_param_match_poll_a_t2(poll_a_prev, poll_a_current);
+            }
+            break;
+        case NCI_RF_INTERFACE_ISO_DEP:
+            /* ISO-DEP Type 4A */
+            if(mode_param_prev && mode_param_current) {
+                const NciModeParamPollA* poll_a_prev = &mode_param_prev->poll_a;
+                const NciModeParamPollA* poll_a_current =
+                    &mode_param_current->poll_a;
+
+                return mode_param_match_poll_a(poll_a_prev, poll_a_current);
+            }
+            break;
```
And this one could/should be much shorter:
```c
        case NCI_RF_INTERFACE_ISO_DEP: /* ISO-DEP Type 4A */
            return mode_param_match_poll_a(&mp1->poll_a, &mp2->poll_a);
```

ikamaletdinov replied:
I have refactored that part of code. But unfortunately, parameters could be NULL, but I have moved that check one layer above (before switch

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+        break;
+    case NCI_MODE_PASSIVE_POLL_B:
+        switch (ntf->rf_intf) {
+        case NCI_RF_INTERFACE_ISO_DEP:
+            /* ISO-DEP Type 4B */
+            if(mode_param_prev && mode_param_current) {
+                const NciModeParamPollB* poll_b_prev = &mode_param_prev->poll_b;
+                const NciModeParamPollB* poll_b_current =
+                    &mode_param_current->poll_b;
+
+                return mode_param_match_poll_b(poll_b_prev, poll_b_current);
+            }
+            break;
```
```c
        case NCI_RF_INTERFACE_ISO_DEP: /* ISO-DEP Type 4B */
            return mode_param_match_poll_b(&mp1->poll_b, &mp2->poll_b);
```

ikamaletdinov replied:
I have refactored that part of code. But unfortunately, parameters could be NULL, but I have moved that check one layer above (before switch

monich commented on src/nci_adapter.c:
```diff
@@ -142,9 +293,7 @@ nci_adapter_intf_info_matches(
-        info->mode_param.size == ntf->mode_param_len &&
-        (!ntf->mode_param_len || !memcmp(info->mode_param.bytes,
-        ntf->mode_param_bytes, ntf->mode_param_len)) &&
+        nci_adapter_info_mode_params_matches (info, ntf) &&
```
Kill the space before `(` please

monich commented on src/nci_adapter.c:
```diff
@@ -430,7 +579,7 @@ nci_adapter_nci_intf_activated(
         }
         if (!tag) {
-            nfc_adapter_add_other_tag(NFC_ADAPTER(self), self->target);
+            nfc_adapter_add_other_tag2(NFC_ADAPTER(self), self->target, NULL);
```
This isn't any better than it was before. The point of the deprecation warning was to make sure that the actual poll parameters are passed in (in cases when they make sense), not NULL.

ikamaletdinov replied:
Thanks a lot, I have fixed that

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
             info->activation_param.bytes = NULL;
         }
+
+        if(ntf->mode_param) {
```
No need for this check, `nci_util_copy_mode_param` checks it and returns NULL.

ikamaletdinov replied:
Indeed, thanks

monich commented on src/nci_adapter.c:
```diff
@@ -88,11 +90,20 @@ G_DEFINE_ABSTRACT_TYPE(NciAdapter, nci_adapter, NFC_TYPE_ADAPTER)
+    if (x && x->mode_param_parsed)                  \
+        g_free(x->mode_param_parsed);               \
+    g_free(x);                                      \
+} while (0);
```
Let's kill this macro and just do
```c
            g_free(priv->active_intf->mode_param_parsed);
            g_free(priv->active_intf);
```
in the only place where it's actually used.

ikamaletdinov replied:
done

monich replied:
But the macro is still alive!

ikamaletdinov replied:
yeap, but now it is dead :)

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +138,151 @@ nci_adapter_intf_info_new(
+    case NCI_MODE_PASSIVE_LISTEN_15693:
+        break;
+    }
+    /* otherwise full match is expected */
```
**C**apitalize the first word

ikamaletdinov commented:
@monich Thanks a lot for the review, I have resolved most of your comments and I have added some refactoring as an idea, please make a review again. Thanks in advance

ikamaletdinov committed:
[nciplugin] Fix deprecated function warning. JB#51201

Usage of `nfc_adapter_add_other_tag` was deprecated. Use
`nfc_adapter_add_other_tag2` instead with correct parameters.

ikamaletdinov committed:
[nciplugin] Fix wrong compare logic. JB#51201

Comparison logic during reactivation was wrong for t2 and t4 tags.
This commit changes behaviour to desired.

monich commented on src/nci_adapter.c:
```diff
@@ -430,7 +554,8 @@ nci_adapter_nci_intf_activated(
         if (!tag) {
-            nfc_adapter_add_other_tag(NFC_ADAPTER(self), self->target);
+            nfc_adapter_add_other_tag2(NFC_ADAPTER(self), self->target,
+                nfc_tag_param(tag));
```
Wait, but if `tag` is NULL then wouldn't `nfc_tag_param(NULL)` just always return `NULL` and that's it? The point of the deprecation warning was that poll parameters should still be provided whenever possible, even if the exact type of tag is unknown or unsupported.

ikamaletdinov replied:
I have changed that logic, please review, thanks

monich commented on src/nci_adapter.c:
```diff
@@ -397,6 +521,46 @@ nci_adapter_create_known_tag(
+NfcTag*
+nci_adapter_create_other_tag(
+    NciAdapter* self,
+    const NciIntfActivationNtf* ntf)
```
I would prefer it this way:
```c
static
const NfcParamPoll*
nci_adapter_get_mode_param(
    NfcParamPoll* poll,
    const NciIntfActivationNtf* ntf)
```
and then the caller would pass that to `nfc_adapter_add_other_tag2()`.

monich replied:
Actually, it's pretty much a matter of taste, you can keep it as it if you like. But in this case don't return any value (since it's not being used) and call `NFC_ADAPTER(self)` once at the beginning of the function. That would do it.

ikamaletdinov replied:
Thanks, I like your approach, lets try again, I have refactored a little bit

monich commented on src/nci_adapter.c:
```diff
@@ -397,6 +521,46 @@ nci_adapter_create_known_tag(
+    /* Figure out what kind of target we are dealing with */
+    switch (ntf->mode) {
+    case NCI_MODE_PASSIVE_POLL_A:
+        if (mp) {     /* Caller didn't check it for NULL */
```
The common practice in this code is to add `/* Caller checked it for NULL */` kind of comment where a pointer **is not** checked for NULL. Checking for NULL is considered a normal practice not requiring any comments.

ikamaletdinov replied:
Thanks, fixed

monich commented on src/nci_adapter.c:
```diff
@@ -397,6 +521,46 @@ nci_adapter_create_known_tag(
+        }
+        break;
+    case NCI_MODE_PASSIVE_POLL_B:
+        if (mp) {     /* Caller didn't check it for NULL */
```
And here too.

ikamaletdinov replied:
thanks, fixed

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +131,132 @@ nci_adapter_intf_info_new(
+            case NCI_RF_INTERFACE_FRAME:
+                /* Type 2 Tag */
+                return mode_param_match_poll_a_t2(&mp1->poll_a, &mp2->poll_a);
+                break;
```
oops, unnecessary break

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +131,132 @@ nci_adapter_intf_info_new(
+            case NCI_RF_INTERFACE_ISO_DEP:
+                /* ISO-DEP Type 4A */
+                return mode_param_match_poll_a(&mp1->poll_a, &mp2->poll_a);
+                break;
```
another one

monich commented on src/nci_adapter.c:
```diff
@@ -127,11 +131,132 @@ nci_adapter_intf_info_new(
+            case NCI_RF_INTERFACE_ISO_DEP:
+                /* ISO-DEP Type 4B */
+                return mode_param_match_poll_b(&mp1->poll_b, &mp2->poll_b);
+                break;
```
and one more!

monich commented on src/nci_adapter.c:
```diff
@@ -430,7 +588,9 @@ nci_adapter_nci_intf_activated(
         }
         if (!tag) {
-            nfc_adapter_add_other_tag(NFC_ADAPTER(self), self->target);
+            NfcParamPoll poll;
```
Empty line after the declaration pls! That's the last one, I promise))

ikamaletdinov replied:
thanks, I have fixed :+1:

monich reviewed (approved):
LGTM now!
